hardware/is_camera/src/camera.cc

```diff
@@ -306,19 +341,67 @@ namespace is_camera {
         bayer_img_msg_buffer_[i]->step = kImageWidth;
         bayer_img_msg_buffer_[i]->data.resize(kImageWidth * kImageHeight);
       }
-  }
-  // Enable or disable the feature timer
-  bool CameraNodelet::EnableBayerService(ff_msgs::SetBool::Request & req,
-                     ff_msgs::SetBool::Response & res) {
-    if (req.enable && !bayer_enable_) {
-      EnableBayer();
-    } else if (!req.enable && bayer_enable_) {
+    } else if (!enable && bayer_enable_) {
       bayer_pub_.shutdown();
+    } else {
+      return;
+    }
+    enable = bayer_enable_;
+  }
+
+  // Timer that periodically changes camera exposure based on the captured image's histogram
+  void CameraNodelet::AutoExposure() {
+    // If images not being captured, do not adjust exposure
+    if (!thread_running_ || ((pub_.getNumSubscribers() == 0) && (getNumBayerSubscribers() == 0))) return;
+
+    // Get last generated image from buffer (index incremented after image completed)
+    sensor_msgs::ImagePtr& grey_image = img_msg_buffer_[(img_msg_buffer_idx_ - 1) % kImageMsgBuffer];
+
+    cv::Mat input(kImageHeight, kImageWidth,
+                     cv::DataType<uint8_t>::type,
+                     &(grey_image->data[0]),
+                     kImageWidth);
+
+    // Calculate Histogram
+    cv::Mat hist;
+    int hist_size = 5;
+    float range[] = {0, 256};  // the upper boundary is exclusive
+    const float* hist_range[] = { range };
+    bool uniform = true, accumulate = false;
+    cv::calcHist(&input, 1, 0, cv::Mat(), hist, 1, &hist_size, hist_range, uniform, accumulate);
+
+    // Calculate mean sample value
+    double mean_sample_value = 0;
+
+    for (int i = 0; i < hist_size; ++i) {
+      mean_sample_value += hist.at<float>(i) * (i+1);
+    }
+    mean_sample_value /= (kImageHeight * kImageWidth);
+
+    // Control Auto exposure with a PI controller
```

Comment:
I would tentatively recommend replacing the PI controller with the strategy implemented at [1]. This is from an open source driver of the same type of camera we are using (TIS).

The code there is quite complex in terms of having heuristics for trading off exposure time, gain, and iris setting, but if you ignore the existence of multiple exposure parameters and only think about exposure time, it's quite simple:
- If the mean is within 5% of the desired value, don't adjust
- If the mean is further than 5% from the desired value, adjust by 5% toward the desired value.
- (And clip the parameter to the valid range.)

I think by using division instead of subtraction to compare with the desired value, this approach already copes with gamma correction better. Hard coding 5% makes it less tunable than the PI controller, but the flip side is that we can hope the algorithm's presence here means it was thoroughly tested by the vendor and they didn't find much need to tune it.

As we figured out this afternoon, a feature that might be good as a second pass improvement (but not until we have more data for tuning) is to switch the auto exposure algorithm to increase the gain instead of the exposure time once the exposure time reaches a pre-specified maximum value related to motion blur.

[1] https://github.com/TheImagingSource/tiscamera/blob/master/src/algorithms/auto_exposure.cpp

Reply:
It seems that you can reduce their algorithm (second bullet) to:

exposure=exposure*(2/3+1/3(reference/actual))